### DISCUSSION TITLE
Compute Desmos Expressions from JavaScript

### DISCUSSION THREAD

**Discussion by radian628** [DesModder/DesModder#857] [Ideas]
What if you could have JavaScript embedded into a graph that generates Desmos expressions at compile-time? This would be useful for lookup tables, 3D model importing, and a whole host of other common(ish) things. 

This _would_ be a huge security issue if not for the fact that we can just use [iframes](https://www.figma.com/blog/how-we-built-the-figma-plugin-system/) to sandbox everything. It didn't work for Figma because they had to pass around massive amounts of data in thousands of roundtrips per second. However, I don't think we have that issue. If that *does* ever become a problem, we might be able to try one of their other approaches.

**Comment by radian628** [comment]
More thoughts: Iframes still have the issue of allowing network requests, thus opening up the possibility of IP grabbers. Another option would be to use something like [QuickJS](https://www.npmjs.com/package/quickjs-emscripten), which is just a JS interpreter that we can run in a sandbox.

There's also the approach outlined in the Figma article of using the Realms shim, a polyfill for a TC39 proposal for "Realms", which appear to be sandboxed JavaScript execution contexts. However, there was [a security issue](https://www.figma.com/blog/an-update-on-plugin-security/) discovered in Realms, so it might be better to just go with the quickjs interpreter if anything:

**Comment by radian628** [comment]
Update: Seems like Desmos doesn't really care about IP grabbers anyway (since it loads in any image from any site) so an iframe is probably good enough.

**Comment by jared-hughes** [comment]
Sounds like sandboxed iframes would work fine. The performance is no problem. Regarding the design, I suggest a "play" button to insert the iframe (so it's always opt-in) and a "stop" button to ensure the iframe can be removed at any time.
